src/Controller/BackendCsvImportController.php

```diff
@@ -0,0 +1,356 @@
+            } catch (\RuntimeException $e) {
+                $request->getSession()->getFlashBag()->add($e->getMessage());
+
+                return new RedirectResponse($request->getUri(), 303);
+            }
+
+            $this->connection->update(
+                $table,
+                [$field => serialize($data)],
+                ['id' => $id]
+            );
+
+            $response = new RedirectResponse($this->getBackUrl($request));
+            $response->headers->setCookie(new Cookie('BE_PAGE_OFFSET', 0, 0));
+
+            return $response;
+        }
+
+        return new Response($template->parse());
+    }
+
+    /**
+     * Creates the CSV import template.
+     *
+     * @param Request    $request
+     * @param FileUpload $uploader
+     * @param bool       $allowLinebreak
+     *
+     * @return Template|\stdClass
+     */
+    private function prepareTemplate(Request $request, FileUpload $uploader, $allowLinebreak = false)
+    {
+        /** @var BackendTemplate|\stdClass $template */
+        $template = new BackendTemplate('be_csv_import');
+
+        $template->formId      = $this->getFormId($request);
+        $template->backUrl     = $this->getBackUrl($request);
+        $template->action      = $request->getRequestUri();
+        $template->fileMaxSize = $this->framework->getAdapter(Config::class)->get('maxFileSize');
+        $template->messages    = $request->getSession()->getFlashBag()->all();
+        $template->uploader    = $uploader->generateMarkup();
+        $template->separators  = $this->getSeparators($allowLinebreak);
+        $template->submitLabel = $GLOBALS['TL_LANG']['MSC']['apply'][0];
+
+        return $template;
+    }
+
+    /**
+     * Returns an array of data from imported CSV files.
+     *
+     * @param FileUpload $uploader
+     * @param string     $separator
+     * @param callable   $callback
+     *
+     * @return array
+     */
+    private function fetchData(FileUpload $uploader, $separator, callable $callback)
+    {
+        $data      = [];
+        $files     = $this->getFiles($uploader);
+        $delimiter = $this->getDelimiter($separator);
+
+        foreach ($files as $file) {
+            $fp = fopen($file, 'rb');
+
+            while (($row = fgetcsv($fp, 0, $delimiter)) !== false) {
+                $data = $callback($data, $row);
+            }
+        }
+
+        return $data;
+    }
+
+    /**
+     * Returns the form ID for the template.
+     *
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function getFormId(Request $request)
+    {
+        return 'tl_csv_import_'.$request->query->get('key');
+    }
+
+    /**
+     * Returns the back button and redirect URL.
+     *
+     * @param Request $request
+     *
+     * @return string
+     */
+    private function getBackUrl(Request $request)
+    {
+        return str_replace('&key='.$request->query->get('key'), '', $request->getRequestUri());
```

Comment:
There will never be only `key` as parameter, because that would not work in the backend. I do understand the "do it perfectly", but I think that's over-engineered for this case. Very likely the request query parameters to not match what's actually in the URL…